app/src/main/java/com/eveningoutpost/dexdrip/BluetoothScan.java

```diff
@@ -155,16 +172,93 @@ public void run() {
+                    @Override
+                    public void run() {
+                        for (ScanResult result: results) {
+                            mLeDeviceListAdapter.addDevice(result.getDevice());
+                        }
+                        mLeDeviceListAdapter.notifyDataSetChanged();
+                    }
+                });
+            }
+
+            @Override
+            public void onScanResult(int callbackType, ScanResult result) {
+                final BluetoothDevice device = result.getDevice();
+                runOnUiThread(new Runnable() {
+                    @Override
+                    public void run() {
+                        mLeDeviceListAdapter.addDevice(device);
+                        mLeDeviceListAdapter.notifyDataSetChanged();
+                    }
+                });
+            }
+        };
+    }
+
+    @TargetApi(21)
+    private void scanLeDeviceLollipop(final boolean enable) {
+        if (enable) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+                lollipopScanner = bluetooth_adapter.getBluetoothLeScanner();
+            }
+
+            Log.d(TAG, "Starting scanner 21");
+            // Stops scanning after a pre-defined scan period.
+            mHandler.postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    is_scanning = false;
+                    lollipopScanner.stopScan(mScanCallback);
+                    invalidateOptionsMenu();
+                }
+            }, SCAN_PERIOD);
+            ScanSettings settings = new ScanSettings.Builder()
+                    .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
+                    .build();
+            is_scanning = true;
+            lollipopScanner.startScan(null, settings, mScanCallback);
+        } else {
+            is_scanning = false;
+            lollipopScanner.stopScan(mScanCallback);
+        }
+        invalidateOptionsMenu();
+    }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, Intent data) {
+        IntentResult scanResult = IntentIntegrator.parseActivityResult(requestCode, resultCode, data);
+        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
+        if (scanResult == null || scanResult.getContents() == null) {
+            return;
+        }
+        if (scanResult.getFormatName().equals("CODE_128")) {
+            Log.d(TAG, "Setting serial number to: " + scanResult.getContents());
+            prefs.edit().putString("share_key", scanResult.getContents()).apply();
+            returnToHome();
+        }
+    }
+
     @Override
     protected void onListItemClick(ListView l, View v, int position, long id) {
         Log.d(TAG, "Item Clicked");
         final BluetoothDevice device = mLeDeviceListAdapter.getDevice(position);
-        if (device == null) return;
+        if (device == null || device.getName() == null) return;
```

Comment:
If we are returning here should I remove the code that lists nameless devices as 'unknown device' and just have it skip them entirely?